### ISSUE TITLE
apostrophe-favicons is not quite uploadfs-compliant

### ISSUE THREAD

**Issue by kafkiano** [apostrophecms/apostrophe-favicons#2]
Hi

I implemented apostrophe-favicons in my project fullpage.xyz and im trying to fix an error which is probably caused because i am using wasabi for the live version.

**Its working on local mashine with the following settings:**
```
    'apostrophe-favicons': {},
    'apostrophe-favicons-global': {
      destinationDir: '/fav/',
      tempDir: 'temp',
      faviconConfig : {
        icons: {
          windows: false
        }
      }
    },
```

----

**These settings i used for connecting wasabi storage:**
```
heroku config:set APOS_S3_BUCKET=YOUR-bucket-name
heroku config:set APOS_S3_SECRET=YOUR-access-key
heroku config:set APOS_S3_KEY=YOUR-secret-key
heroku config:set APOS_S3_ENDPOINT=s3.wasabisys.com
```

----

**its working well for all attachments but i get some strange path error when favicon-progress starts:**

```
2018-06-09T17:55:32.705231+00:00 app[web.1]: { Error: ENOENT: no such file or directory, open '/app/publichttp://fullpage.s3.wasabisys.com/attachments/cji7p97z9000r0kw8g9gk0ddc-internet.png'
2018-06-09T17:55:32.705248+00:00 app[web.1]: errno: -2,
2018-06-09T17:55:32.705252+00:00 app[web.1]: code: 'ENOENT',
2018-06-09T17:55:32.705253+00:00 app[web.1]: syscall: 'open',
2018-06-09T17:55:32.705256+00:00 app[web.1]: path: '/app/publichttp://fullpage.s3.wasabisys.com/attachments/cji7p97z9000r0kw8g9gk0ddc-internet.png' }
2018-06-09T17:55:32.708133+00:00 app[web.1]: (node:20) UnhandledPromiseRejectionWarning: TypeError: Cannot read property 'images' of undefined
2018-06-09T17:55:32.708136+00:00 app[web.1]: at /app/node_modules/apostrophe-favicons/lib/modules/apostrophe-favicons-global/index.js:223:37
2018-06-09T17:55:32.708137+00:00 app[web.1]: at <anonymous>
2018-06-09T17:55:32.708211+00:00 app[web.1]: (node:20) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 1)
2018-06-09T17:55:32.708280+00:00 app[web.1]: (node:20) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
```

----

**I have found the following settings in apostrophe-favicons-global/index.js:**

 ```
// Retrieve the file path to the favicon image
        function getImage(callback) {
          return images.find(req, {
            '_id': piece.aposFavicon.items[0].pieceIds[0]
          }).toObject(function (err, image) {
            if (err) {
              console.log(err);
            }

            let path = self.apos.rootDir + '/public' + attachments.url(image.attachment, {
              uploadfsPath: false,
              size: 'original'
            });

            return callback(null, path);

          });
        };
```

----

**How can i adjust them to make them fit into my wasabi url i would be really glad if you could help the there shortly.**

**Comment by boutell** [comment]
@stuartromanek this code:

```
let path = self.apos.rootDir + '/public' + attachments.url(image.attachment, {
             uploadfsPath: false,
             size: 'original'
           });
```

Isn't quite uploadfs-ready. You need to use `uploadfs.copyOut` to copy it to a local temporary file.

**Comment by stuartromanek** [comment]
@felixlberg I just published a fix for this, v1.0.11, can you implement this in your environment and let us know if it is fixed?

**Comment by kafkiano** [comment]
I've tested now seems that retrieving the file path ist working now.

`2018-06-11T19:00:06.237918+00:00 heroku[router]: at=info method=POST path="/modules/apostrophe-global/favicon-progress" host=fullpage.xyz request_id=52461200-4fe0-46ee-bdcf-b034f86d1739 fwd="95.121.51.30" dyno=web.1 connect=1ms service=21285ms status=200 bytes=396 protocol=http`

but when it creates the temporary directory the following error happens.
```
2018-06-11T19:00:06.224198+00:00 app[web.1]: { Error: ENOENT: no such file or directory, mkdir '/app//app/temp'
2018-06-11T19:00:06.224210+00:00 app[web.1]: at Object.fs.mkdirSync (fs.js:885:18)
2018-06-11T19:00:06.224213+00:00 app[web.1]: at /app/node_modules/apostrophe-favicons/lib/modules/apostrophe-favicons-global/index.js:219:16
2018-06-11T19:00:06.224215+00:00 app[web.1]: at /app/node_modules/favicons/dist/index.js:57:14
2018-06-11T19:00:06.224217+00:00 app[web.1]: at <anonymous>
2018-06-11T19:00:06.224218+00:00 app[web.1]: at process._tickDomainCallback (internal/process/next_tick.js:228:7)
2018-06-11T19:00:06.224220+00:00 app[web.1]: errno: -2,
2018-06-11T19:00:06.224222+00:00 app[web.1]: code: 'ENOENT',
2018-06-11T19:00:06.224224+00:00 app[web.1]: syscall: 'mkdir',
2018-06-11T19:00:06.224225+00:00 app[web.1]: path: '/app//app/temp' }
2018-06-11T19:00:06.227216+00:00 app[web.1]: (node:20) UnhandledPromiseRejectionWarning: Error: ENOENT: no such file or directory, mkdir '/app//app/temp'
2018-06-11T19:00:06.227219+00:00 app[web.1]: at Object.fs.mkdirSync (fs.js:885:18)
2018-06-11T19:00:06.227221+00:00 app[web.1]: at /app/node_modules/apostrophe-favicons/lib/modules/apostrophe-favicons-global/index.js:219:16
2018-06-11T19:00:06.227223+00:00 app[web.1]: at <anonymous>
2018-06-11T19:00:06.227224+00:00 app[web.1]: at process._tickDomainCallback (internal/process/next_tick.js:228:7)
```
locally it's working perfect with these settings.

**Comment by stuartromanek** [comment]
@felixlberg can you try v1.0.12 in your environment?

Also, I removed the option to provide your own temp dir path, uploadfs really just needs to figure that out on its own

**Comment by kafkiano** [comment]
I've tested now in live environment. I have also removed temp dir settings in app.js. Like you said they are removed from apostrophe-favicons-global anyway. My settings are like this now:
```

    'apostrophe-favicons-global': {
      destinationDir: '/fav/',
      faviconConfig : {
        icons: {
          windows: false
        }
      }
    }
```
With these settings it works now in my environment.

I would consider it as closed thx very much for the effort...

**Comment by kafkiano** [comment]
Unfortunately it just worked once, if i try to render a new image to favicon it seems he crashes now at retrieving the path of the previous created temp dir.

```
2018-06-11T22:59:49.410889+00:00 app[web.1]: { Error: ENOENT: no such file or directory, mkdir '/app/data/temp/uploadfs//app/data/temp/uploadfs//app/data/temp/uploadfs/tempAposFavicons'
2018-06-11T22:59:49.410911+00:00 app[web.1]: at Object.fs.mkdirSync (fs.js:885:18)
2018-06-11T22:59:49.410914+00:00 app[web.1]: at /app/node_modules/apostrophe-favicons/lib/modules/apostrophe-favicons-global/index.js:199:16
2018-06-11T22:59:49.410916+00:00 app[web.1]: at /app/node_modules/favicons/dist/index.js:57:14
2018-06-11T22:59:49.410918+00:00 app[web.1]: at <anonymous>
2018-06-11T22:59:49.410920+00:00 app[web.1]: at process._tickDomainCallback (internal/process/next_tick.js:228:7)
2018-06-11T22:59:49.410923+00:00 app[web.1]: errno: -2,
2018-06-11T22:59:49.410925+00:00 app[web.1]: code: 'ENOENT',
2018-06-11T22:59:49.410927+00:00 app[web.1]: syscall: 'mkdir',
2018-06-11T22:59:49.410929+00:00 app[web.1]: path: '/app/data/temp/uploadfs//app/data/temp/uploadfs//app/data/temp/uploadfs/tempAposFavicons' }
2018-06-11T22:59:49.411346+00:00 app[web.1]: (node:20) UnhandledPromiseRejectionWarning: Error: ENOENT: no such file or directory, mkdir '/app/data/temp/uploadfs//app/data/temp/uploadfs//app/data/temp/uploadfs/tempAposFavicons'
2018-06-11T22:59:49.411350+00:00 app[web.1]: at Object.fs.mkdirSync (fs.js:885:18)
2018-06-11T22:59:49.411352+00:00 app[web.1]: at /app/node_modules/apostrophe-favicons/lib/modules/apostrophe-favicons-global/index.js:199:16
2018-06-11T22:59:49.411354+00:00 app[web.1]: at <anonymous>
2018-06-11T22:59:49.411356+00:00 app[web.1]: at process._tickDomainCallback (internal/process/next_tick.js:228:7)
```
Update:
I think the problem won't last long because of the none persistent storage of heroku. or does the temporary directory now gets created trough uploadfs directly in the s3 bucket ?
sry i didn't understood the uploadfs mechanism fully right now....

I will test tomorrow further to reproduce this error

**Comment by kafkiano** [comment]
i can confirm that path is created twice if you change the image like above.

**Comment by stuartromanek** [comment]
@felixlberg was able to get this figured and fixed in `v1.0.13` .. not sure why it didn't crop up before this, I couldn't find it having anything to do with using cloud storage but construction of the temp path was getting cached by the server and then re-appended to, causing crazyness. It is constructed with that in mind now
